File: internal/db/models/backup.go
/**
 * Backup database models.
 */

package models

import (
	"time"

	"gorm.io/gorm"
)

// BackupStatus represents the status of a backup job.
type BackupStatus string

const (
	BackupStatusPending   BackupStatus = "pending"
	BackupStatusRunning   BackupStatus = "running"
	BackupStatusCompleted BackupStatus = "completed"
	BackupStatusFailed    BackupStatus = "failed"
	BackupStatusCanceled  BackupStatus = "canceled"
)

// BackupType represents the type of backup.
type BackupType string

const (
	BackupTypeFull     BackupType = "full"
	BackupTypeFiles    BackupType = "files"
	BackupTypeDatabase BackupType = "database"
)

// BackupJob represents a backup job record.
type BackupJob struct {
	ID             string         `gorm:"primaryKey;size:26" json:"id"`
	TenantID       string         `gorm:"size:26;index;not null" json:"tenant_id"`
	AccountID      string         `gorm:"size:26;index" json:"account_id"`
	DomainID       string         `gorm:"size:26;index" json:"domain_id"`
	Type           BackupType     `gorm:"size:20;not null" json:"type"`
	Status         BackupStatus   `gorm:"size:20;not null;index" json:"status"`
	Name           string         `gorm:"size:255" json:"name"`
	Description    string         `gorm:"size:500" json:"description"`
	StorageBackend string         `gorm:"size:50;default:local" json:"storage_backend"`
	StoragePath    string         `gorm:"size:500" json:"storage_path"`
	FileSize       int64          `gorm:"default:0" json:"file_size"`
	FileCount      int            `gorm:"default:0" json:"file_count"`
	Checksum       string         `gorm:"size:128" json:"checksum"`
	ChecksumAlgo   string         `gorm:"size:20;default:sha256" json:"checksum_algo"`
	RetentionDays  int            `gorm:"default:30" json:"retention_days"`
	ExpiresAt      *time.Time     `gorm:"index" json:"expires_at"`
	ErrorMessage   string         `gorm:"size:1000" json:"error_message,omitempty"`
	ScheduledAt    *time.Time     `json:"scheduled_at,omitempty"`
	StartedAt      *time.Time     `json:"started_at,omitempty"`
	CompletedAt    *time.Time     `json:"completed_at,omitempty"`
	CreatedBy      string         `gorm:"size:26" json:"created_by"`
	CreatedAt      time.Time      `json:"created_at"`
	UpdatedAt      time.Time      `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName returns the table name for BackupJob.
func (BackupJob) TableName() string {
	return "backup_jobs"
}

// BackupFile represents a file within a backup.
type BackupFile struct {
	ID           string         `gorm:"primaryKey;size:26" json:"id"`
	BackupJobID  string         `gorm:"size:26;index;not null" json:"backup_job_id"`
	OriginalPath string         `gorm:"size:1000;not null" json:"original_path"`
	ArchivePath  string         `gorm:"size:1000;not null" json:"archive_path"`
	Size         int64          `gorm:"default:0" json:"size"`
	Checksum     string         `gorm:"size:128" json:"checksum"`
	IsDirectory  bool           `gorm:"default:false" json:"is_directory"`
	CreatedAt    time.Time      `json:"created_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName returns the table name for BackupFile.
func (BackupFile) TableName() string {
	return "backup_files"
}

// RestorePoint represents a restore operation record.
type RestorePoint struct {
	ID              string         `gorm:"primaryKey;size:26" json:"id"`
	TenantID        string         `gorm:"size:26;index;not null" json:"tenant_id"`
	AccountID       string         `gorm:"size:26;index" json:"account_id"`
	DomainID        string         `gorm:"size:26;index" json:"domain_id"`
	BackupJobID     string         `gorm:"size:26;index;not null" json:"backup_job_id"`
	Status          BackupStatus   `gorm:"size:20;not null" json:"status"`
	StagingDir      string         `gorm:"size:500" json:"staging_dir"`
	TargetDir       string         `gorm:"size:500" json:"target_dir"`
	FilesRestored   int            `gorm:"default:0" json:"files_restored"`
	BytesRestored   int64          `gorm:"default:0" json:"bytes_restored"`
	RollbackEnabled bool           `gorm:"default:true" json:"rollback_enabled"`
	RollbackUsed    bool           `gorm:"default:false" json:"rollback_used"`
	ErrorMessage    string         `gorm:"size:1000" json:"error_message,omitempty"`
	StartedAt       *time.Time     `json:"started_at,omitempty"`
	CompletedAt     *time.Time     `json:"completed_at,omitempty"`
	CreatedBy       string         `gorm:"size:26" json:"created_by"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName returns the table name for RestorePoint.
func (RestorePoint) TableName() string {
	return "restore_points"
}

// BackupConfig represents storage backend configuration.
type BackupConfig struct {
	ID             string    `gorm:"primaryKey;size:26" json:"id"`
	TenantID       string    `gorm:"size:26;index;not null" json:"tenant_id"`
	Name           string    `gorm:"size:100;not null" json:"name"`
	Backend        string    `gorm:"size:50;not null" json:"backend"` // local, s3, minio, wasabi
	Endpoint       string    `gorm:"size:255" json:"endpoint"`
	Bucket         string    `gorm:"size:255" json:"bucket"`
	Region         string    `gorm:"size:50" json:"region"`
	AccessKeyID    string    `gorm:"size:255" json:"access_key_id,omitempty"`
	SecretAccessKey string   `gorm:"size:500" json:"secret_access_key,omitempty"` // encrypted
	PathPrefix     string    `gorm:"size:255;default:backups" json:"path_prefix"`
	IsDefault      bool      `gorm:"default:false" json:"is_default"`
	RetentionDays  int       `gorm:"default:30" json:"retention_days"`
	CreatedAt      time.Time `json:"created_at"`
	UpdatedAt      time.Time `json:"updated_at"`
	DeletedAt      gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName returns the table name for BackupConfig.
func (BackupConfig) TableName() string {
	return "backup_configs"
}

// BackupSchedule represents a scheduled backup job.
type BackupSchedule struct {
	ID           string         `gorm:"primaryKey;size:26" json:"id"`
	TenantID     string         `gorm:"size:26;index;not null" json:"tenant_id"`
	AccountID    string         `gorm:"size:26;index" json:"account_id"`
	DomainID     string         `gorm:"size:26;index" json:"domain_id"`
	Name         string         `gorm:"size:100;not null" json:"name"`
	Description  string         `gorm:"size:500" json:"description"`
	BackupType   BackupType     `gorm:"size:20;not null" json:"backup_type"`
	CronExpr     string         `gorm:"size:100;not null" json:"cron_expr"`
	RetentionDays int           `gorm:"default:30" json:"retention_days"`
	StorageBackend string       `gorm:"size:50;default:local" json:"storage_backend"`
	IsEnabled    bool           `gorm:"default:true" json:"is_enabled"`
	LastRunAt    *time.Time     `json:"last_run_at,omitempty"`
	NextRunAt    *time.Time     `json:"next_run_at,omitempty"`
	CreatedBy    string         `gorm:"size:26" json:"created_by"`
	CreatedAt    time.Time      `json:"created_at"`
	UpdatedAt    time.Time      `json:"updated_at"`
	DeletedAt    gorm.DeletedAt `gorm:"index" json:"-"`
}

// TableName returns the table name for BackupSchedule.
func (BackupSchedule) TableName() string {
	return "backup_schedules"
}

File: frontend/src/lib/api/backup.ts
/**
 * Backup API client.
 */

import { request } from "./client";

// Types
export type BackupStatus = "pending" | "running" | "completed" | "failed" | "canceled";
export type BackupType = "full" | "files" | "database";

export interface BackupJob {
  id: string;
  tenant_id: string;
  account_id?: string;
  domain_id?: string;
  type: BackupType;
  status: BackupStatus;
  name?: string;
  description?: string;
  storage_backend: string;
  storage_path?: string;
  file_size: number;
  file_count: number;
  checksum?: string;
  checksum_algo: string;
  retention_days: number;
  expires_at?: string;
  error_message?: string;
  scheduled_at?: string;
  started_at?: string;
  completed_at?: string;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface BackupFile {
  id: string;
  backup_job_id: string;
  original_path: string;
  archive_path: string;
  size: number;
  checksum: string;
  is_directory: boolean;
  created_at: string;
}

export interface RestorePoint {
  id: string;
  tenant_id: string;
  account_id?: string;
  domain_id?: string;
  backup_job_id: string;
  status: BackupStatus;
  staging_dir?: string;
  target_dir?: string;
  files_restored: number;
  bytes_restored: number;
  rollback_enabled: boolean;
  rollback_used: boolean;
  error_message?: string;
  started_at?: string;
  completed_at?: string;
  created_by: string;
  created_at: string;
}

export interface BackupSchedule {
  id: string;
  tenant_id: string;
  account_id?: string;
  domain_id?: string;
  name: string;
  description?: string;
  backup_type: BackupType;
  cron_expr: string;
  retention_days: number;
  storage_backend: string;
  is_enabled: boolean;
  last_run_at?: string;
  next_run_at?: string;
  created_by: string;
  created_at: string;
}

export interface BackupStats {
  total_backups: number;
  active_backups: number;
  failed_backups: number;
  total_storage_mb: number;
}

export interface CreateBackupRequest {
  account_id?: string;
  domain_id?: string;
  type: BackupType;
  name?: string;
  description?: string;
  storage_backend?: string;
  retention_days?: number;
  source_path?: string;
}

export interface RestoreBackupRequest {
  target_dir: string;
  files_to_restore?: string[];
  rollback_enabled?: boolean;
}

export interface CreateScheduleRequest {
  account_id?: string;
  domain_id?: string;
  name: string;
  description?: string;
  backup_type: BackupType;
  cron_expr: string;
  retention_days?: number;
  storage_backend?: string;
}

export interface ListBackupsResponse {
  backups: BackupJob[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

// API Functions
export const backupKeys = {
  all: ["backups"] as const,
  lists: () => [...backupKeys.all, "list"] as const,
  list: (filters?: Record<string, string>) => [...backupKeys.lists(), filters] as const,
  stats: () => [...backupKeys.all, "stats"] as const,
  details: () => [...backupKeys.all, "detail"] as const,
  detail: (id: string) => [...backupKeys.details(), id] as const,
  files: (id: string) => [...backupKeys.detail(id), "files"] as const,
  restores: (id: string) => [...backupKeys.detail(id), "restores"] as const,
  schedules: () => [...backupKeys.all, "schedules"] as const,
};

export async function listBackups(params?: {
  page?: number;
  page_size?: number;
  type?: string;
  status?: string;
}): Promise<ListBackupsResponse> {
  const searchParams = new URLSearchParams();
  if (params?.page) searchParams.set("page", String(params.page));
  if (params?.page_size) searchParams.set("page_size", String(params.page_size));
  if (params?.type) searchParams.set("type", params.type);
  if (params?.status) searchParams.set("status", params.status);

  return request<ListBackupsResponse>(`/api/v1/backups?${searchParams.toString()}`);
}

export async function getBackup(id: string): Promise<BackupJob> {
  return request<BackupJob>(`/api/v1/backups/${id}`);
}

export async function createBackup(data: CreateBackupRequest): Promise<BackupJob> {
  return request<BackupJob>("/api/v1/backups", { method: "POST", body: data });
}

export async function deleteBackup(id: string): Promise<void> {
  await request(`/api/v1/backups/${id}`, { method: "DELETE" });
}

export async function getBackupFiles(id: string): Promise<{ backup_id: string; files: BackupFile[] }> {
  return request<{ backup_id: string; files: BackupFile[] }>(`/api/v1/backups/${id}/files`);
}

export async function restoreBackup(id: string, data: RestoreBackupRequest): Promise<RestorePoint> {
  return request<RestorePoint>(`/api/v1/backups/${id}/restore`, { method: "POST", body: data });
}

export async function getRestorePoints(id: string): Promise<{ backup_id: string; restore_points: RestorePoint[] }> {
  return request<{ backup_id: string; restore_points: RestorePoint[] }>(`/api/v1/backups/${id}/restores`);
}

export async function getBackupStats(): Promise<BackupStats> {
  return request<BackupStats>("/api/v1/backups/stats");
}

// Schedule API
export async function createSchedule(data: CreateScheduleRequest): Promise<BackupSchedule> {
  return request<BackupSchedule>("/api/v1/backups/schedules", { method: "POST", body: data });
}

export async function listSchedules(): Promise<{ schedules: BackupSchedule[] }> {
  return request<{ schedules: BackupSchedule[] }>("/api/v1/backups/schedules");
}

export async function deleteSchedule(id: string): Promise<void> {
  await request(`/api/v1/backups/schedules/${id}`, { method: "DELETE" });
}